src/mini_s3.erl

```diff
@@ -115,327 +123,148 @@ manual_start() ->
     application:start(ssl),
     application:start(inets).
 
--spec new(string(), string()) -> config().
-
-new(AccessKeyID, SecretAccessKey) ->
-    #config{
-     access_key_id=AccessKeyID,
-     secret_access_key=SecretAccessKey}.
-
--spec new(string(), string(), string()) -> config().
-
-new(AccessKeyID, SecretAccessKey, Host) ->
-    #config{
-     access_key_id=AccessKeyID,
-     secret_access_key=SecretAccessKey,
-     s3_url=Host}.
-
--spec new(string(), string(), string(), bucket_access_type()) -> config().
+-spec parse_ipv_host_domain(string()) -> {pos_integer(), list(), string()}.
+parse_ipv_host_domain(Host0) ->
+    case string:lexemes(Host0, "[]") of
+        % ipv4
+        [Host0] ->
+            Domain0 = "",
+            {4, Host0, Domain0};
+        % ipv6
+        [Scheme0,    Domain0, Port0] -> {6, lists:flatten([Scheme0,    $\r, Port0]), Domain0};
+        ["http://",  Domain0       ] -> {6, lists:flatten(["http://",  $\r       ]), Domain0};
+        ["https://", Domain0       ] -> {6, lists:flatten(["https://", $\r       ]), Domain0};
+        [            Domain0, Port0] -> {6, lists:flatten([            $\r, Port0]), Domain0};
+        [            Domain0       ] -> {6,               [            $\r       ] , Domain0}
+    end.
 
+-spec new(string() | binary(), string() | binary(), string()) -> aws_config().
+new(AccessKeyID, SecretAccessKey, Host0) ->
+    % chef-server crams scheme://host:port all into into Host; erlcloud wants them separate.
+    % Assume:
+    %   Host   == scheme://domain:port | scheme://domain | domain:port | domain
+    %   scheme == http | https
+
+    % ipv4/6 detection
+    {Ipv, Host, Domain0} = parse_ipv_host_domain(Host0),
```

Comment:
Have we looked into using this: http://erlang.org/doc//man/uri_string.html I wonder if it might get us most of what we want? But, you've written enough tests that this isn't my biggest concern, so no worries if not.

Reply:
After reading your comment, I did quite a bit of work converting over to `uri_string`.  The issue was, it does not correctly parse `"host:port"`, whether host is ipv4, or ipv6.  I was able to code up a correction in the ipv4 case without adding too much code or complexity.  But if we correct for the ipv6 case, the resulting amount of code/complexity will be more than if we simply go with what we have.  Your call, and I'll do whatever you recommend.

Reply:
EDIT:  "Hold the phone, Chuck."  I may have an acceptable solution.

Reply:
Done.